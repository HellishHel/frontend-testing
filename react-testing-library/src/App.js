import './App.css';
import {useEffect, useState} from "react";
import {Users} from "./users/Users";

function App() {
    const [data, setData] = useState(null)
    const [toggle, setToggle] = useState(false)
    const [value, setValue] = useState("")

    const onClick = () => setToggle((prevState) => !prevState)

    useEffect(() => {
        setTimeout(() => {
            setData({})
        }, 1000)
    }, []);

  return (
    <div>
        <h1 data-testid="value-elem">{value}</h1>
        {toggle && <div data-testid="toggle-elem">toggle</div>}
        {data && <div style={{color: "red"}}>data</div>}
      <h1>Hello world</h1>
      <button onClick={onClick} data-testid="toggle-btn">Click me</button>
      <input onChange={(event) => setValue(event.target.value)}
             type="text"
             placeholder="input value"/>
        <Users/>
    </div>
  );
}

export default App;
