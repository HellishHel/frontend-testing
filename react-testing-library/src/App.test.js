import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';
import userEvent from "@testing-library/user-event";

describe("Test app", () => {
  test('get', () => {
    render(<App />);
    const helloWorldElem = screen.getByText(/hello world/i);
    const btn = screen.getByRole("button")
    const input = screen.getByPlaceholderText(/input value/i)

    expect(helloWorldElem).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  });

  test('query', () => {
    render(<App />);
    const helloWorldElem2 = screen.queryByText(/hello2/i);

    expect(helloWorldElem2).toBeNull();
  });

  test('find', async () => {
    render(<App />);

    screen.debug()

    const elem = await screen.findByText(/data/i);
    expect(elem).toBeInTheDocument();
    expect(elem).toHaveStyle({color: "red"})

    screen.debug()
  });

  test('click event', async () => {
    render(<App />);

    const btn = screen.getByTestId("toggle-btn");

    expect(screen.queryByTestId("toggle-elem")).toBeNull()
    fireEvent.click(btn)
    expect(screen.queryByTestId("toggle-elem")).toBeInTheDocument()
    fireEvent.click(btn)
    expect(screen.queryByTestId("toggle-elem")).toBeNull()
  });

  test('input event', async () => {
    render(<App />);

    const input = screen.getByPlaceholderText(/input value/i)

    expect(screen.getByTestId("value-elem")).toContainHTML("")
    // искуственное событие

    // fireEvent.input(input, {
    //   target: {
    //     value: "123"
    //   }
    // })

    // близко к поведению пользователя, обрабатываются события нажатия клавиш и т.п.
    userEvent.type(input, "123")

    expect(screen.getByTestId("value-elem")).toContainHTML("123")
  });
})
