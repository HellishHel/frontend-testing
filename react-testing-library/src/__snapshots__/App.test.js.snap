// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test app get 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`Test app renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;
